File: 033 OOPS in C++/007 Inheritance3.cpp
#include <bits/stdc++.h>

class Car {
    
    public:
    
    string name;
    string color;
    int engineHP; //engine horse power
    
    Car() {
        
    }
    
    Car(string name, string color, int engineHP) {
        this->name = name;
        this->color = color;
        this->engineHP = engineHP; 
    }
    
    void drive() {
        cout<<name<<" Car is driving\n";
    }
    
};

class FlyingCar : public Car {
    
    public:
    FlyingCar() {
        
    }
    
    FlyingCar(string name,string color, int engineHP) {
        this->name = name;
        this->color = color;
        this->engineHP = engineHP;
    }
    
    void fly() {
        cout<<name<<" Car is flying\n";
    }
};

class HydroCar : public Car {
    public:
    HydroCar() {
        
    }
    
    HydroCar(string name,string color, int engineHP) {
        this->name = name;
        this->color = color;
        this->engineHP = engineHP;
    }
    
    void swim() {
        cout<<name<<" Car is travelling on the surface of the water\n";
    }  
};

int main() {
    
    FlyingCar fc("SuperFly","Black",5000);
    fc.drive();
    fc.fly();
    
    HydroCar hc("SuperHydro","Black",5000);
    hc.drive();
    hc.swim();
}

File: 032 OOPS in Java/007 Inheritance3.java
import java.util.*;

class Car {
    String name;
    String color;
    int engineHP; //engine Horse Power
    
    Car(String name, String color,int engineHP) {
        this.name = name;
        this.color = color;
        this.engineHP = engineHP;
    }
    
    public void drive() {
       System.out.println(name + " Car is driving");
    }
    
    public void reverse() {
        System.out.println(name + " Car is in reverse mode");
    }
}

class FlyingCar extends Car {
    
    FlyingCar(String name, String color, int engineHP) {
        super(name,color,engineHP);
    }
    
    public void fly() {
        System.out.println("Flying Car is flying");
    }
}

class HydroCar extends Car {
    
    HydroCar(String name, String color, int engineHP) {
        super(name,color,engineHP);
    }
    
    public void swim() {
        System.out.println("Hydro car is driving on the surface of the water");
    }
}

public class Main {
    public static void main(String[] args) {
        FlyingCar fc = new FlyingCar("SuperFly","Black",5000);
        fc.drive();
        fc.reverse();
        fc.fly();
        
        HydroCar hc = new HydroCar("SuperSwim","Black",5000);
        hc.drive();
        hc.reverse();
        hc.swim();
    }
}
